tensorstore/kvstore/s3/s3_request_builder.h

```diff
@@ -0,0 +1,135 @@
+// Copyright 2020 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TENSORSTORE_KVSTORE_S3_REQUEST_BUILDER_H_
+#define TENSORSTORE_KVSTORE_S3_REQUEST_BUILDER_H_
+
+/// \file
+/// S3 Request Builder
+
+#include <optional>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+
+#include "absl/log/absl_check.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "tensorstore/internal/ascii_utils.h"
+#include "tensorstore/internal/http/http_request.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/kvstore/byte_range.h"
+#include "tensorstore/internal/digest/sha256.h"
+#include "tensorstore/kvstore/s3/validate.h"
+#include "tensorstore/util/result.h"
+
+#include "absl/strings/str_split.h"
+#include "absl/time/time.h"
+
+using ::tensorstore::internal_http::HttpRequest;
+using ::tensorstore::internal::IntToHexDigit;
+using ::tensorstore::internal::ParseGenericUri;
+using ::tensorstore::internal::SHA256Digester;
+
+namespace tensorstore {
+namespace internal_storage_s3 {
+
+std::string UriEncode(std::string_view src);
+std::string UriObjectKeyEncode(std::string_view src);
+
+/// Size of HMAC (size of SHA256 digest).
+constexpr static size_t kHmacSize = 32;
+
+void ComputeHmac(std::string_view key, std::string_view message, unsigned char (&hmac)[kHmacSize]) {
+    unsigned int md_len = kHmacSize;
+    // Computing HMAC should never fail.
+    ABSL_CHECK(HMAC(EVP_sha256(),
+                    reinterpret_cast<const unsigned char*>(key.data()),
+                    key.size(),
+                    reinterpret_cast<const unsigned char*>(message.data()),
+                    message.size(), hmac, &md_len) &&
+               md_len == kHmacSize);
+}
+
+void ComputeHmac(unsigned char (&key)[kHmacSize], std::string_view message, unsigned char (&hmac)[kHmacSize]){
+    unsigned int md_len = kHmacSize;
+    // Computing HMAC should never fail.
+    ABSL_CHECK(HMAC(EVP_sha256(), key, kHmacSize,
+                    reinterpret_cast<const unsigned char*>(message.data()),
+                    message.size(), hmac, &md_len) &&
+               md_len == kHmacSize);
+}
+
+
+
+class S3RequestBuilder {
```

Comment:
This could actually inherit from HTTPRequestBuilder, and be the main interface for S3. Then you wouldn't need to duplicate the Add...Header functions.

Reply:
Code re-use would be great, but I think the best we can do here is proxying an `HttpRequestBuilderr` in `S3RequestBuilder`. With basic static or dynamic inheritance, all inherited methods would return a `HttpRequestBuilder &` as in the following example:

```cpp
#include <cassert>
#include <iostream>
#include <numeric>
#include <string>
#include <vector>

class request {
public:
 request(const std::vector<std::string>& headers) : value_(std::accumulate(headers.begin(), headers.end(), std::string(""))) {}
 std::string value_;
};

class base {
public:
  base & AddHeader(std::string header) { headers.push_back(header); return *this; }
  request BuildRequest() { return request(this->headers); } 
  std::vector<std::string> headers;
};

class derived : public base {
public:
  request BuildRequest(std::string data) { headers.push_back(data); return request(this->headers); }
};


int main()
{
    auto request = derived().AddHeader("hello world").BuildRequest(/* if derived, we could pass through data */);
    std::cout << request.value_ << std::endl;
    assert(request.value_ == "hello world");
}

};
```

With encapsulation we'd end up proxying all the `HttpRequestBuilder` calls but these can be inlineable in the header. Similarly, we could proxy functions like `AddRangeHeader`.

Reply:
Ah actually, I remember why I didn't use inheritance here (and encapsulation would also struggle here). The base AddQueryParameter uses `PercentEncodeUriComponent` to encode key/value pairs, but we actually want to use the AWS `UriEncode` here. So I ended up re-implementing the `HttpRequestBuilder`.

Perhaps passing in a URI Encode functor into the Request Builder class might work?

I'll think about this some more, but let me know if there's some other approach.

Reply:
Seems fine.
```c++

class HttpRequestBuilder {
 public:
  explicit HttpRequestBuilder(absl::FunctionRef<std::string(std::string_view)> encode_functor)
  ...
};
```

Reply:
> Seems fine.
> 
> ```c++
> class HttpRequestBuilder {
>  public:
>   explicit HttpRequestBuilder(absl::FunctionRef<std::string(std::string_view)> encode_functor)
>   ...
> };
> ```

Thanks for pointing that out, will use it.

Also just to telegraph, `S3RequestBuilder` will encapsulate a `HttpRequestBuilder` object.